### ISSUE TITLE
Does multi-gpu tensorflow code can be simply migrate to horovod ?

### ISSUE THREAD

**Issue by zh794390558** [uber/horovod#102]
Does multi-gpu tensorflow code can be simply migrate to horovod with opt.compute_gradients() and opt.apply_gradients()?

**Comment by alsrgv** [comment]
Sure, you can simply squeeze this code in between:

```python
gradients = ...

# Average gradients using Horovod
averaged_gradients = []
for grad, var in gradients:
    if grad is not None:
        averaged_gradients.append((hvd.allreduce(grad), var))
    else:
        averaged_gradients.append((None, var))

apply_op = opt.apply_gradients(averaged_graduents)
...
```

**Comment by weixsong** [comment]
@alsrgv , does this mean that:
I could allocate 8 GPU on 2 node, 4 GPU for each node.
And in each node, I could use the multiple-gpu gradients average code, and then use your code to do distributed computation between these two node with 8 GPU?

**Comment by alsrgv** [comment]
@weixsong, yes, you can.  However, traditional tower-based averaging is typically slower and requires more code.

**Comment by weixsong** [comment]
@alsrgv , more code is not an issue for me. But slower maybe the issue.
Then you mean if launch one process for each GPU would be much faster than the tower way? for example 2 towers, 8 gpu for each tower?

My understanding if I use tower based gradient averaging, then gradient communication between each node would be much less, only two copy of gradient need to be send over network. By my understanding, this way could reduce the network transfer cost. 
But, maybe MPI could be more efficiency than my understanding.

**Comment by alsrgv** [comment]
@weixsong, you can do an experiment. I saw on two very different hardware configurations that ring-allreduce within a server was faster than tower-based approach. It would be interesting if it's not the case on your system.

Also, never underestimate the power of simpler code. :-)

**Comment by zh794390558** [comment]
@alsrgv   If I have 4 gpus, 2 nodes, 2 gpus per node. Then I need 2 process which can see 2 gpus per process, right?  Then use averaged gradients to do all-reduce for 2 processes? I need one process to see all 2 gpus on one node ?

**Comment by alsrgv** [comment]
@zh794390558, if you want to use code [like this](https://github.com/uber/horovod/issues/102#issuecomment-346135571), then yes.

**Comment by ppwwyyxx** [comment]
@alsrgv On a DGX1 running `tensorflow/benchmarks` with resnet50, I actually saw that gradient averaging within tensorflow is slightly faster than horovod (using your [PR](https://github.com/tensorflow/benchmarks/pull/73)).

**Comment by alsrgv** [comment]
@ppwwyyx, that's an interesting result!  Can you specify which flags were you using and what scaling factors are you seeing?

On my system with 8xP100 on SXM2 training ResNet-50 with batch size 64 per GPU, I get 229.86 img/sec on a single GPU, 1519.87 img/sec on 8xGPU (6.61x) with towers and 1715.18 img/sec on 8xGPU with Horovod and NCCL (7.46x).

**Comment by ppwwyyxx** [comment]
Oh, the test was about a month ago when I saw your PR. Just now I tested again and I saw better performance for horovod (horovod-1759 vs replicated-1643). Nice work!
